### PR TITLE
unit tests for jkube-kit/common dir

### LINKED ISSUE #207: Increase Unit test coverage for jkube-kit/common (achieve 45% code coverage)
# Description
Increase current unit test coverage for the following directory https://github.com/eclipse/jkube/tree/master/jkube-kit/common 

# Acceptance Criteria
- [ ] 45% Code coverage for the [specified module](https://sonarcloud.io/component_measures?id=jkubeio_jkube&metric=coverage&selected=jkubeio_jkube%3Ajkube-kit%2Fcommon%2Fsrc)
- [ ] It should cover complex scenario
- [ ] Scenario test should be passed, otherwise, issues will re-reported

/kind user-story

### PR DESCRIPTION
**Pull request by anandrkskd**
## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
 - [x] I have read the [contributing guidelines](https://www.eclipse.org/jkube/contributing)
 - [x] I signed-off my commit with a user that has signed the [Eclipse Contributor Agreement](https://www.eclipse.org/legal/ECA.php)
 - [ ] ~I Added [CHANGELOG](../CHANGELOG.md) entry~
 - [x] I have implemented unit tests to cover my changes
 - [ ] ~I have updated the [documentation](../kubernetes-maven-plugin/doc) accordingly~
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=jkubeio_jkube) report
 - [ ] ~I tested my code in Kubernetes~
 - [ ] ~I tested my code in OpenShift~

### CODE REVIEW

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 1]
```diff
@@ -62,4 +197,12 @@ public void testLoadTimestampShouldLoadFromFile() throws Exception {
         // Then
         assertThat(timestamp, equalTo(new Date(1445385600000L)));
     }
+
+    @Test
+    public  void testIsWindows(){
+        //When
+        boolean result = isWindows();
+        //Then
+        assertEquals(false,result);
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:206]: I think this can be simplified to `assertFalse(isWindows())`

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 2]
```diff
@@ -23,15 +23,150 @@
...
+    }
+
+    @Test
+    public void testSplitOnSpaceWithEscape(){
+        //Given
+        String strings = "hello world";
+        //when
+        String[] result = EnvUtil.splitOnSpaceWithEscape(strings);
+        //then
+        //System.out.println("result:"+ Arrays.toString(result));
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:140]: remove this line

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 3]
```diff
@@ -23,15 +23,150 @@
...
+        //List<String[]> result2 = EnvUtil.splitOnLastColon(list2);
+        //Then
+        assertEquals(2,result1.get(0).length);
+        //System.out.println(result2.get(0).length);
+        //assertEquals("",result2.get(0));
+    }
+
+    /*
+    @Test
+    public void testSplitAtCommasAndTrim(){
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:109]: Please remove dead code

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 4]
```diff
@@ -23,15 +23,150 @@
...
+        List<String> list1 = new ArrayList<>();
+        //List<String> list2 = new ArrayList<>();
+        list1.add("element1:element2");
+        //list2.add(null);
+        //When
+        List<String[]> result1 = EnvUtil.splitOnLastColon(list1);
+        //List<String[]> result2 = EnvUtil.splitOnLastColon(list2);
+        //Then
+        assertEquals(2,result1.get(0).length);
+        //System.out.println(result2.get(0).length);
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:103]: Please remove dead code

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 5]
```diff
@@ -23,15 +23,150 @@
...
+        assertNull(EnvUtil.extractLargerVersion(null,null));
+        assertEquals(result2,versionA);
+        assertEquals(result3,versionB);
+        assertEquals(result4,versionA);
+        assertEquals(result5,versionA);
+        assertEquals("4.0.0.1",result6);
+    }
+
+    @Test
+    public void testGreaterOrEqualsVersion(){
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:77]:
```suggestion
    public void testGreaterOrEqualsVersion() {
```

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 6]
```diff
@@ -23,15 +23,150 @@
...
+        String result3 = EnvUtil.extractLargerVersion(null,versionB);
+        String result4 = EnvUtil.extractLargerVersion(versionA,versionB);
+        String result5 = EnvUtil.extractLargerVersion(versionB,versionA);
+        String result6 = EnvUtil.extractLargerVersion("4.0.0.1","4.0.0");
+        //Then
+        assertNull(EnvUtil.extractLargerVersion(null,null));
+        assertEquals(result2,versionA);
+        assertEquals(result3,versionB);
+        assertEquals(result4,versionA);
+        assertEquals(result5,versionA);
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:72]: I think `assertEquals()` expects expected item as the first argument and actual arguments as the second. You need to swap these

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java (hunk 1 above)
Comment (anandrkskd) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:206]: Understood!

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java (hunk 3 above)
Comment (anandrkskd) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:109]: I was not sure how to test that portion of code so I commented it.

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java (hunk 6 above)
Comment (anandrkskd) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:72]: Understood

**Comment by manusa** [comment]
Just as a generic guideline, you only want to test a single thing (unit) in a unit test, so if the test fails you know exactly what caused the failure.
I see many tests implemented here which try to test several things at a time.

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 7]
```diff
@@ -23,15 +23,320 @@
...
+        assertEquals("http://0.0.0.0:2375", result2);
+    }
+
+    @Test
+    public void testExtractLargerVersionWhenBothNull(){
+        //Given
+        // Both version are null
+        //When
+        //Then
+        assertNull(EnvUtil.extractLargerVersion(null,null));
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:76]: Please, remove the comments:
```suggestion
        assertNull(EnvUtil.extractLargerVersion(null,null));
```
Or replace by  declarative code:
```suggestion
        //Given
        var v1 = null;
        var v2 = null;
        //When
        var result = EnvUtil.extractLargerVersion(v1, v2);
        //Then
        assertNull(result);
```

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 8]
```diff
@@ -23,15 +23,320 @@
...
+        // Then
+        assertEquals("http://0.0.0.0:2375", result2);
+    }
+
+    @Test
+    public void testExtractLargerVersionWhenBothNull(){
+        //Given
+        // Both version are null
+        //When
+        //Then
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:75]: nit, maybe you don't need to write `// Given... when... Then` when test statements are one-liners.

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 9]
```diff
@@ -23,15 +23,320 @@
...
+        // When
+        String result2 = EnvUtil.convertTcpToHttpUrl(urlWithHttpPort);
+        // Then
+        assertEquals("http://0.0.0.0:2375", result2);
+    }
+
+    @Test
+    public void testExtractLargerVersionWhenBothNull(){
+        //Given
+        // Both version are null
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:73]: I don't think you need to add this comment since test name indicates the same

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 10]
```diff
@@ -23,15 +23,320 @@
...
+        //Then
+        assertNull(EnvUtil.extractLargerVersion(null,null));
+    }
+    @Test
+    public void testExtractLargerVersionWhenBIsNull(){
+        //Given
+        String versionA = "4.0.2";
+        String versionB = "3.1.1.0";
+        //When
+        String result = EnvUtil.extractLargerVersion(versionA,null);
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:84]: Seems like a copy-paste leftover
```suggestion
        String versionB = null;
        //When
        String result = EnvUtil.extractLargerVersion(versionA, versionB);
```

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 11]
```diff
@@ -23,15 +23,320 @@
...
+        //Then
+        assertEquals( versionA, result);
+    }
+    @Test
+    public void testExtractLargerVersionWhenAIsNull(){
+        //Given
+        String versionA = "4.0.2";
+        String versionB = "3.1.1.0";
+        //When
+        String result = EnvUtil.extractLargerVersion(null,versionB);
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:94]: Seems like a copy-paste leftover
```suggestion
        //Given
        String versionA = null;
        String versionB = "3.1.1.0";
        //When
        String result = EnvUtil.extractLargerVersion(versionA, versionB);
```

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 12]
```diff
@@ -23,15 +23,320 @@
...
+        //Then
+        assertEquals( versionB, result);
+    }
+    @Test
+    public void testExtractLargerVersionWhenAisGreater(){
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:99]: This test is redundant/duplicate of testExtractLargerVersion()

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 13]
```diff
@@ -23,15 +23,320 @@
...
+        //Then
+        assertEquals(2 ,result.size());
+        assertEquals("value",result.get("name"));
+    }
+
+    @Test
+    public void testFormatDurationTill() throws InterruptedException {
+        long startTime = System.currentTimeMillis();
+        // Wait for 200 milliseconds
+        Thread.sleep(200L);
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:257]: This is very bad, try not to use this in tests ever. I don't even see the purpose of this sleep/wait in the test.

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 14]
```diff
@@ -62,4 +367,10 @@ public void testLoadTimestampShouldLoadFromFile() throws Exception {
         // Then
         assertThat(timestamp, equalTo(new Date(1445385600000L)));
     }
+
+    @Test
+    public  void testIsWindows(){
+        //Then
+        assertEquals(false,isWindows());
```
Comment (manusa) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:374]: Platform dependent test. Try to make it platform independent and provide two tests, one that verifies it is windows and one that verifies it's not.

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java (hunk 13 above)
Comment (rohanKanojia) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:257]: I suggested him to use `Thread.sleep()` since the test is measuring time duration since a specified value

**Comment by manusa** [comment]
Very good progress, I'm only missing some edge-cases 

e.g. `EnvUtil.extractLargerVersion("1005-sasd@|@~½!", "asdasd|@#~¬{!!");`


Also, please remember to squash your commits

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java (hunk 13 above)
Comment (manusa) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:257]: Never use `Thread.sleep` in tests!

For this test, I don't even understand why we are using `system.currentTimeMillis();`, but if this is really necessary, then you can simply replace by:
```suggestion
long startTime = System.currentTimeMillis() - 200L;
```

Some quick sources with reasons why (although there are maaaany bad reasons to do so)
- http://www.kyleblaney.com/junit-best-practices#:~:text=Do%20not%20use%20Thread.&text=When%20a%20unit%20test%20uses,the%20production%20code%20is%20broken. 
- https://laurentkempe.com/2012/11/20/Using-ThreadSleep-in-unit-test-A-good-idea/

Comment (rohanKanojia) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:257]: ohk :neutral_face: .

File: jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java [hunk 15]
```diff
@@ -16,33 +16,332 @@
...
+        //When
+        boolean result3 = EnvUtil.greaterOrEqualsVersion(versionB,versionA);
+        //Then
+        assertFalse(result3);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    @Ignore
+// TODO: Remove when implementation is fixed
+    public void testGreaterOrEqualsVersionCornerCase() {
```
Comment (rohanKanojia) [inline jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:137]: Why are we ignoring this test?

Comment (manusa) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:137]: image

Comment (rohanKanojia) [reply jkube-kit/common/src/test/java/org/eclipse/jkube/kit/common/util/EnvUtilTest.java:137]: ah, okay.
